engine/runtime/src/main/java/org/enso/compiler/SuggestionsCache.java

```diff
@@ -0,0 +1,151 @@
+      throws CacheException {
+    if (obj instanceof Suggestions suggestions) {
+      return new CachedSuggestions(libraryName, suggestions);
+    } else {
+      throw new CacheException("Expected SuggestionsCache.Suggestions, got " + obj.getClass());
+    }
+  }
+
+  @Override
+  protected Optional<Metadata> metadataFromBytes(byte[] bytes) {
+    var maybeJsonString = new String(bytes, Cache.metadataCharset);
+    var mapper = new ObjectMapper();
+    try {
+      return Optional.of(mapper.readValue(maybeJsonString, SuggestionsCache.Metadata.class));
+    } catch (JsonProcessingException e) {
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  protected Optional<String> computeDigest(CachedSuggestions entry, TruffleLogger logger) {
+    var digest = messageDigest();
+    entry.getSuggestions().getSuggestions().forEach(suggestion -> {
+      ByteBuffer bytes = ByteBuffer.allocate(Integer.BYTES).putInt(suggestion.hashCode());
+      digest.update(bytes);
+    });
+    return Optional.of(Hex.toHexString(digest.digest()));
+  }
+
+  @Override
+  protected Optional<String> computeDigestFromSource(EnsoContext context, TruffleLogger logger) {
+    return context
+        .getPackageRepository()
+        .getPackageForLibraryJava(libraryName)
+        .map(pkg -> computeDigestOfLibrarySources(pkg.listSourcesJava(), logger));
+  }
+
+
+  @Override
+  protected Optional<Roots> getCacheRoots(EnsoContext context) {
+    return context.getPackageRepository().getPackageForLibraryJava(libraryName).map(pkg -> {
+      var bindingsCacheRoot = pkg.getSuggestionsCacheRootForPackage(Info.ensoVersion());
+      var localCacheRoot = bindingsCacheRoot.resolve(libraryName.namespace());
+      var distribution = context.getDistributionManager();
+      var pathSegments = CollectionConverters.ListHasAsScala(Arrays.asList(
+              pkg.namespace(),
+              pkg.name(),
+              pkg.config().version(),
+              Info.ensoVersion(),
+              libraryName.namespace())
+      ).asScala();
+      var path = distribution.LocallyInstalledDirectories().irCacheDirectory()
+              .resolve(pathSegments.mkString("/"));
+      var globalCacheRoot = context.getTruffleFile(path.toFile());
+      return new Cache.Roots(localCacheRoot, globalCacheRoot);
+    });
+  }
+
+  @Override
+  protected Object extractObjectToSerialize(CachedSuggestions entry) {
+    return entry.getSuggestions();
+  }
+
+  // Suggestions class is not a record because of a Frgaal bug leading to invalid compilation error.
+  final static class Suggestions implements Serializable {
+
+    private final List<Suggestion> suggestions;
+
+    public Suggestions(List<Suggestion> suggestions)  {
+      this.suggestions = suggestions;
+    }
+
+    public List<Suggestion> getSuggestions() {
+      return suggestions;
+    }
+  }
+
+  // CachedSuggestions class is not a record because of a Frgaal bug leading to invalid compilation error.
```

Comment:
What is the error? Maybe updating to new version of frgaal would fix the problem.

Reply:
Yes, this is what I was seeing as well and the reason why in other cases I don't use the record. Thanks for reporting it with frgaal